upload/views/helpers.py

```diff
@@ -45,3 +50,45 @@ def get_repository_from_string(
         service=service, repo_identifier=repo_identifier
     )
     return repository
+
+
+def get_repository_and_owner_from_slug_and_commit(
+    slug: str, commitid: str
+) -> tuple[Repository | None, Owner | None]:
+    if "::::" not in slug:
+        return None, None
+
+    owner_username, repository_name = slug.rsplit("::::", 1)
+
+    # formatting for GL subgroups
+    if ":::" in owner_username:
+        owner_username = owner_username.replace(":::", ":")
+
+    matching_repositories = Repository.objects.filter(
+        author__username=owner_username, name=repository_name
+    ).select_related("author")
+    if matching_repositories.count() == 1:
+        log.info(
+            "get_repository_and_owner_from_slug_and_commit success",
+            extra=dict(slug=slug),
+        )
+        repository = matching_repositories.first()
+        return repository, repository.author
+    else:
```

Comment:
is there ever a case where matching_repositories is empty? I think that may be the case if the user has never synced repos

Reply:
yes - I think it's edge case but possible to have no matches in this repo query. I'm not kicking them out at this point because this query being empty means the secondary query with the commitid won't find any matches, but it means I am doing the secondary query knowing that it will fail. I can change that

Reply:
yeah i think it's unlikely the repo won't exist, since users would typically need to get the upload token to upload and they'd get that from the UI, i think it's fine that it crashes later during the commitid step